[Bug]: "bad math long / long long settings" when consuming wolfssl on arm32 android

### Contact Details

_No response_

### Version

5.8.4

### Description

While trying to update port curl in vcpkg, I found this build error for vcpkg's arm-neon-android triplet:

~~~
FAILED: [code=1] lib/CMakeFiles/libcurl_static.dir/md5.c.o 
/android-ndk-r29/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi28 --sysroot=/android-ndk-r29/toolchains/llvm/prebuilt/linux-x86_64/sysroot -DBUILDING_LIBCURL -DCURL_HIDDEN_SYMBOLS -DECC_SHAMIR -DECC_TIMING_RESISTANT -DGCM_TABLE_4BIT -DHAVE_AESGCM -DHAVE_AES_KEYWRAP -DHAVE_ALPN -DHAVE_CERTIFICATE_STATUS_REQUEST -DHAVE_CERTIFICATE_STATUS_REQUEST_V2 -DHAVE_CHACHA -DHAVE_CONFIG_H -DHAVE_CRL -DHAVE_DH_DEFAULT_PARAMS -DHAVE_ECC -DHAVE_ENCRYPT_THEN_MAC -DHAVE_EXTENDED_MASTER -DHAVE_FFDHE_2048 -DHAVE_HASHDRBG -DHAVE_HKDF -DHAVE_HPKE -DHAVE_MAX_FRAGMENT -DHAVE_OCSP -DHAVE_ONE_TIME_AUTH -DHAVE_PKCS7 -DHAVE_POLY1305 -DHAVE_PTHREAD -DHAVE_SNI -DHAVE_SUPPORTED_CURVES -DHAVE_THREAD_LS -DHAVE_TLS_EXTENSIONS -DHAVE_TRUNCATED_HMAC -DHAVE_TRUSTED_CA -DHAVE_X963_KDF -DNO_DES3_TLS_SUITES -DNO_DSA -DNO_MD4 -DNO_OLD_TLS -DNO_PSK -DNO_RC4 -DOPENSSL_EXTRA -DTFM_ECC256 -DTFM_TIMING_RESISTANT -DWC_NO_ASYNC_THREADING -DWC_RSA_BLINDING -DWC_RSA_PSS -DWOLFSSL_AES_CFB -DWOLFSSL_AES_DIRECT -DWOLFSSL_ALLOW_ENCODING_CA_FALSE -DWOLFSSL_CERT_EXT -DWOLFSSL_CERT_GEN -DWOLFSSL_CERT_REQ -DWOLFSSL_ENCRYPTED_KEYS -DWOLFSSL_KEY_GEN -DWOLFSSL_NO_SHAKE128 -DWOLFSSL_NO_SHAKE256 -DWOLFSSL_OPTIONS_IGNORE_SYS -DWOLFSSL_PSS_LONG_SALT -DWOLFSSL_SHA384 -DWOLFSSL_SHA512 -DWOLFSSL_SYS_CA_CERTS -DWOLFSSL_TLS13 -DWOLFSSL_USE_ALIGN -DWOLF_CRYPTO_CB -I/mnt/vcpkg-ci/b/curl/src/c-8_19_0-3-dddc4aecab.clean/include -I/mnt/vcpkg-ci/b/curl/arm-neon-android-dbg/lib -I/mnt/vcpkg-ci/b/curl/src/c-8_19_0-3-dddc4aecab.clean/lib -isystem /mnt/vcpkg-ci/installed/arm-neon-android/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -Werror=format-security  -fPIC   -fno-limit-debug-info    -fPIC -Werror-implicit-function-declaration -Wextra -Wall -Wpedantic -Wbad-function-cast -Wconversion -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wno-long-long -Wno-multichar -Wpointer-arith -Wshadow -Wsign-compare -Wundef -Wunused -Wwrite-strings -Waddress -Wattributes -Wcast-align -Wcast-qual -Wdeclaration-after-statement -Wdiv-by-zero -Wempty-body -Wendif-labels -Wfloat-equal -Wformat-security -Wignored-qualifiers -Wmissing-field-initializers -Wmissing-noreturn -Wno-padded -Wno-sign-conversion -Wno-switch-default -Wno-switch-enum -Wno-system-headers -Wold-style-definition -Wredundant-decls -Wstrict-prototypes -Wtype-limits -Wunreachable-code -Wunused-parameter -Wvla -Wconditional-uninitialized -Wno-used-but-marked-unused -Wshift-sign-overflow -Wshorten-64-to-32 -Wformat=2 -Wlanguage-extension-token -Wno-covered-switch-default -Wno-disabled-macro-expansion -Wformat-non-iso -Wenum-conversion -Wmissing-variable-declarations -Wno-documentation-unknown-command -Wsometimes-uninitialized -Wdouble-promotion -Wheader-guard -Wpragmas -Wunused-const-variable -Wcomma -Wassign-enum -Wextra-semi-stmt -Wimplicit-fallthrough -Wxor-used-as-pow -Wcast-function-type -Wreserved-identifier -Wno-reserved-macro-identifier -Wno-unsafe-buffer-usage -Wcast-function-type-strict -Wno-format-signedness -fvisibility=hidden -pthread -MD -MT lib/CMakeFiles/libcurl_static.dir/md5.c.o -MF lib/CMakeFiles/libcurl_static.dir/md5.c.o.d -o lib/CMakeFiles/libcurl_static.dir/md5.c.o -c /mnt/vcpkg-ci/b/curl/src/c-8_19_0-3-dddc4aecab.clean/lib/md5.c
In file included from /mnt/vcpkg-ci/b/curl/src/c-8_19_0-3-dddc4aecab.clean/lib/md5.c:76:
In file included from /mnt/vcpkg-ci/installed/arm-neon-android/include/wolfssl/openssl/md5.h:32:
In file included from /mnt/vcpkg-ci/installed/arm-neon-android/include/wolfssl/wolfcrypt/hash.h:29:
/mnt/vcpkg-ci/installed/arm-neon-android/include/wolfssl/wolfcrypt/types.h:1608:6: error: "bad math long / long long settings"
 1608 |     #error "bad math long / long long settings"
      |      ^
/mnt/vcpkg-ci/installed/arm-neon-android/include/wolfssl/wolfcrypt/types.h:1610:1: error: use of empty enum
 1610 | };
      | ^
2 errors generated.
~~~

There is a successful build of wolfssl, but the installed package doesn't allow a successful build of curl the arm32 android.

### Reproduction steps

I can reproduce this locally with
~~~
vcpkg wolfssl:arm-neon-android  curl[core,wolfssl]:arm-neon-android
~~~


### Relevant log output

```shell

```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -148,17 +148,6 @@ check_type_size("long" SIZEOF_LONG)
 check_type_size("time_t" SIZEOF_TIME_T)
 check_type_size("uintptr_t" HAVE_UINTPTR_T)
 
-# SIZEOF_LONG/SIZEOF_LONG_LONG are exposed to applications through options.h
-# (see cmake/options.h.in) so that application code computes the same
-# CTC_SETTINGS as the library. The in-tree build does not include options.h,
-# so define them here for the library and test programs.
-if(HAVE_SIZEOF_LONG)
-    add_definitions("-DSIZEOF_LONG=${SIZEOF_LONG}")
-endif()
-if(HAVE_SIZEOF_LONG_LONG)
-    add_definitions("-DSIZEOF_LONG_LONG=${SIZEOF_LONG_LONG}")
-endif()
-
 # By default, HAVE___UINT128_T gets defined as TRUE,
 # but we want it as 1.
 if(HAVE___UINT128_T)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -113,6 +113,14 @@ check_include_file("sys/stat.h" HAVE_SYS_STAT_H)
 check_include_file("sys/types.h" HAVE_SYS_TYPES_H)
 check_include_file("unistd.h" HAVE_UNISTD_H)
 
+# types.h depends on HAVE_LIMITS_H, and it is defined in options.h (rather than
+# config.h) so that applications consuming wolfSSL headers see it. The in-tree
+# build, however, is configured through config.h/compile definitions and does
+# not include options.h, so define it here as well.
+if(HAVE_LIMITS_H)
+    add_definitions("-DHAVE_LIMITS_H")
+endif()
+
 include(CheckFunctionExists)
 
 # TODO: Also check if these functions are declared by the
@@ -139,6 +147,17 @@ check_type_size("long long" SIZEOF_LONG_LONG)
 check_type_size("long" SIZEOF_LONG)
 check_type_size("time_t" SIZEOF_TIME_T)
 check_type_size("uintptr_t" HAVE_UINTPTR_T)
+
+# SIZEOF_LONG/SIZEOF_LONG_LONG are exposed to applications through options.h
+# (see cmake/options.h.in) so that application code computes the same
+# CTC_SETTINGS as the library. The in-tree build does not include options.h,
+# so define them here for the library and test programs.
+if(HAVE_SIZEOF_LONG)
+    add_definitions("-DSIZEOF_LONG=${SIZEOF_LONG}")
+endif()
+if(HAVE_SIZEOF_LONG_LONG)
+    add_definitions("-DSIZEOF_LONG_LONG=${SIZEOF_LONG_LONG}")
+endif()
 
 # By default, HAVE___UINT128_T gets defined as TRUE,
 # but we want it as 1.

--- a/cmake/config.in
+++ b/cmake/config.in
@@ -19,9 +19,6 @@
 /* Define to 1 if you have the `gmtime_r' function. */
 #cmakedefine HAVE_GMTIME_R @HAVE_GMTIME_R@
 
-/* Define to 1 if you have the <limits.h> header file. */
-#cmakedefine HAVE_LIMITS_H @HAVE_LIMITS_H@
-
 /* Define to 1 if you have the <pcap/pcap.h> header file. */
 #cmakedefine HAVE_PCAP_PCAP_H @HAVE_PCAP_PCAP_H@
 

--- a/cmake/config.in
+++ b/cmake/config.in
@@ -49,12 +49,6 @@
 /* Define to the full name of this package. */
 #define PACKAGE_NAME "@CMAKE_PROJECT_NAME@"
 
-/* The size of `long', as computed by sizeof. */
-#cmakedefine SIZEOF_LONG @SIZEOF_LONG@
-
-/* The size of `long long', as computed by sizeof. */
-#cmakedefine SIZEOF_LONG_LONG @SIZEOF_LONG_LONG@
-
 /* The size of `time_t', as computed by sizeof. */
 #cmakedefine SIZEOF_TIME_T @SIZEOF_TIME_T@
 

--- a/cmake/config.in
+++ b/cmake/config.in
@@ -49,6 +49,12 @@
 /* Define to the full name of this package. */
 #define PACKAGE_NAME "@CMAKE_PROJECT_NAME@"
 
+/* The size of `long', as computed by sizeof. */
+#cmakedefine SIZEOF_LONG @SIZEOF_LONG@
+
+/* The size of `long long', as computed by sizeof. */
+#cmakedefine SIZEOF_LONG_LONG @SIZEOF_LONG_LONG@
+
 /* The size of `time_t', as computed by sizeof. */
 #cmakedefine SIZEOF_TIME_T @SIZEOF_TIME_T@
 

--- a/cmake/options.h.in
+++ b/cmake/options.h.in
@@ -33,14 +33,14 @@ extern "C" {
 #endif
 
 #ifndef WOLFSSL_OPTIONS_IGNORE_SYS
-/* Since types.h depends on HAVE_LIMITS_H, we must define it in options.h. */
-#undef HAVE_LIMITS_H
-#cmakedefine HAVE_LIMITS_H @HAVE_LIMITS_H@
 #undef _GNU_SOURCE
 #cmakedefine _GNU_SOURCE
 #undef _POSIX_THREADS
 #cmakedefine _POSIX_THREADS
 #endif
+/* Since types.h depends on HAVE_LIMITS_H, we must define it in options.h. */
+#undef HAVE_LIMITS_H
+#cmakedefine HAVE_LIMITS_H @HAVE_LIMITS_H@
 #undef ASIO_USE_WOLFSSL
 #cmakedefine ASIO_USE_WOLFSSL
 #undef BOOST_ASIO_USE_WOLFSSL

--- a/cmake/options.h.in
+++ b/cmake/options.h.in
@@ -33,6 +33,9 @@ extern "C" {
 #endif
 
 #ifndef WOLFSSL_OPTIONS_IGNORE_SYS
+/* Since types.h depends on HAVE_LIMITS_H, we must define it in options.h. */
+#undef HAVE_LIMITS_H
+#cmakedefine HAVE_LIMITS_H @HAVE_LIMITS_H@
 #undef _GNU_SOURCE
 #cmakedefine _GNU_SOURCE
 #undef _POSIX_THREADS

--- a/cmake/options.h.in
+++ b/cmake/options.h.in
@@ -38,14 +38,9 @@ extern "C" {
 #undef _POSIX_THREADS
 #cmakedefine _POSIX_THREADS
 #endif
-/* Since types.h depends on HAVE_LIMITS_H, SIZEOF_LONG and SIZEOF_LONG_LONG,
- * we must define them in options.h. */
+/* Since types.h depends on HAVE_LIMITS_H, we must define it in options.h. */
 #undef HAVE_LIMITS_H
 #cmakedefine HAVE_LIMITS_H @HAVE_LIMITS_H@
-#undef SIZEOF_LONG
-#cmakedefine SIZEOF_LONG @SIZEOF_LONG@
-#undef SIZEOF_LONG_LONG
-#cmakedefine SIZEOF_LONG_LONG @SIZEOF_LONG_LONG@
 #undef ASIO_USE_WOLFSSL
 #cmakedefine ASIO_USE_WOLFSSL
 #undef BOOST_ASIO_USE_WOLFSSL

--- a/cmake/options.h.in
+++ b/cmake/options.h.in
@@ -38,9 +38,14 @@ extern "C" {
 #undef _POSIX_THREADS
 #cmakedefine _POSIX_THREADS
 #endif
-/* Since types.h depends on HAVE_LIMITS_H, we must define it in options.h. */
+/* Since types.h depends on HAVE_LIMITS_H, SIZEOF_LONG and SIZEOF_LONG_LONG,
+ * we must define them in options.h. */
 #undef HAVE_LIMITS_H
 #cmakedefine HAVE_LIMITS_H @HAVE_LIMITS_H@
+#undef SIZEOF_LONG
+#cmakedefine SIZEOF_LONG @SIZEOF_LONG@
+#undef SIZEOF_LONG_LONG
+#cmakedefine SIZEOF_LONG_LONG @SIZEOF_LONG_LONG@
 #undef ASIO_USE_WOLFSSL
 #cmakedefine ASIO_USE_WOLFSSL
 #undef BOOST_ASIO_USE_WOLFSSL

--- a/configure.ac
+++ b/configure.ac
@@ -204,7 +204,9 @@ then
     fi
 fi
 
-AC_CHECK_HEADERS([arpa/inet.h fcntl.h limits.h netdb.h netinet/in.h stddef.h time.h sys/ioctl.h sys/socket.h sys/time.h errno.h sys/un.h ctype.h sys/random.h])
+AC_CHECK_HEADERS([arpa/inet.h fcntl.h netdb.h netinet/in.h stddef.h time.h sys/ioctl.h sys/socket.h sys/time.h errno.h sys/un.h ctype.h sys/random.h])
+# Special case: Since types.h depends on HAVE_LIMITS_H, we must define it in options.h.
+AC_CHECK_HEADER([limits.h], [AM_CPPFLAGS="$AM_CPPFLAGS -DHAVE_LIMITS_H=1"], [])
 AC_CHECK_LIB([network],[socket])
 AC_C_BIGENDIAN
 AC_C___ATOMIC

--- a/configure.ac
+++ b/configure.ac
@@ -207,6 +207,15 @@ fi
 AC_CHECK_HEADERS([arpa/inet.h fcntl.h netdb.h netinet/in.h stddef.h time.h sys/ioctl.h sys/socket.h sys/time.h errno.h sys/un.h ctype.h sys/random.h])
 # Special case: Since types.h depends on HAVE_LIMITS_H, we must define it in options.h.
 AC_CHECK_HEADER([limits.h], [AM_CPPFLAGS="$AM_CPPFLAGS -DHAVE_LIMITS_H=1"], [])
+# Propagate the measured integer sizes (from AC_CHECK_SIZEOF above) into
+# options.h. The library obtains these from config.h, but config.h is not
+# available to applications; without them, application code falls back to the
+# limits.h/architecture heuristics in types.h and can compute a CTC_SETTINGS
+# value that disagrees with the library's CheckRunTimeSettings().
+AS_IF([test "x$ac_cv_sizeof_long" != "x"],
+    [AM_CPPFLAGS="$AM_CPPFLAGS -DSIZEOF_LONG=$ac_cv_sizeof_long"])
+AS_IF([test "x$ac_cv_sizeof_long_long" != "x"],
+    [AM_CPPFLAGS="$AM_CPPFLAGS -DSIZEOF_LONG_LONG=$ac_cv_sizeof_long_long"])
 AC_CHECK_LIB([network],[socket])
 AC_C_BIGENDIAN
 AC_C___ATOMIC

--- a/configure.ac
+++ b/configure.ac
@@ -207,15 +207,6 @@ fi
 AC_CHECK_HEADERS([arpa/inet.h fcntl.h netdb.h netinet/in.h stddef.h time.h sys/ioctl.h sys/socket.h sys/time.h errno.h sys/un.h ctype.h sys/random.h])
 # Special case: Since types.h depends on HAVE_LIMITS_H, we must define it in options.h.
 AC_CHECK_HEADER([limits.h], [AM_CPPFLAGS="$AM_CPPFLAGS -DHAVE_LIMITS_H=1"], [])
-# Propagate the measured integer sizes (from AC_CHECK_SIZEOF above) into
-# options.h. The library obtains these from config.h, but config.h is not
-# available to applications; without them, application code falls back to the
-# limits.h/architecture heuristics in types.h and can compute a CTC_SETTINGS
-# value that disagrees with the library's CheckRunTimeSettings().
-AS_IF([test "x$ac_cv_sizeof_long" != "x"],
-    [AM_CPPFLAGS="$AM_CPPFLAGS -DSIZEOF_LONG=$ac_cv_sizeof_long"])
-AS_IF([test "x$ac_cv_sizeof_long_long" != "x"],
-    [AM_CPPFLAGS="$AM_CPPFLAGS -DSIZEOF_LONG_LONG=$ac_cv_sizeof_long_long"])
 AC_CHECK_LIB([network],[socket])
 AC_C_BIGENDIAN
 AC_C___ATOMIC

--- a/wolfssl/wolfcrypt/types.h
+++ b/wolfssl/wolfcrypt/types.h
@@ -231,6 +231,15 @@ typedef const char wcchar[];
                 (ULONG_MAX == 0xffffffffUL)
             #define SIZEOF_LONG 4
         #endif
+        /* On LP64 (e.g. 64-bit Linux/macOS) long is 8 bytes. Detect it from the
+         * target's own limits.h so CTC_SETTINGS matches the library, which gets
+         * SIZEOF_LONG=8 from config.h. This must be derived per-target (not
+         * baked into options.h), so LLP64 targets such as Windows correctly get
+         * SIZEOF_LONG=4 from the branch above. */
+        #if !defined(SIZEOF_LONG) && defined(ULONG_MAX) && \
+                (ULONG_MAX == 0xffffffffffffffffULL)
+            #define SIZEOF_LONG 8
+        #endif
         #if !defined(SIZEOF_LONG_LONG) && defined(ULLONG_MAX) && \
                 (ULLONG_MAX == 0xffffffffffffffffULL)
             #define SIZEOF_LONG_LONG 8

